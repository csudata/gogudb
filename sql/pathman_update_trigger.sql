\set VERBOSITY terse

SET search_path = 'public','_gogu';
CREATE EXTENSION gogudb;
CREATE SCHEMA test_update_trigger;



/* Partition table by RANGE (NUMERIC) */
CREATE TABLE test_update_trigger.test_range(val NUMERIC NOT NULL, comment TEXT);
INSERT INTO test_update_trigger.test_range SELECT i, i FROM generate_series(1, 100) i;
SELECT create_range_partitions('test_update_trigger.test_range', 'val', 1, 10);
SELECT create_update_triggers('test_update_trigger.test_range');


/* Update values in 1st partition (rows remain there) */
UPDATE test_update_trigger.test_range SET val = 5 WHERE val <= 10;

/* Check values #1 */
SELECT tableoid::REGCLASS, *
FROM test_update_trigger.test_range
WHERE val < 10
ORDER BY comment;

SELECT count(*) FROM test_update_trigger.test_range;


/* Update values in 2nd partition (rows move to 3rd partition) */
UPDATE test_update_trigger.test_range SET val = val + 10 WHERE val > 10 AND val <= 20;

/* Check values #2 */
SELECT tableoid::REGCLASS, *
FROM test_update_trigger.test_range
WHERE val > 20 AND val <= 30
ORDER BY comment;

SELECT count(*) FROM test_update_trigger.test_range;


/* Move single row */
UPDATE test_update_trigger.test_range SET val = 90 WHERE val = 80;

/* Check values #3 */
SELECT tableoid::REGCLASS, *
FROM test_update_trigger.test_range
WHERE val = 90
ORDER BY comment;

SELECT count(*) FROM test_update_trigger.test_range;


/* Move single row (create new partition) */
UPDATE test_update_trigger.test_range SET val = -1 WHERE val = 50;

/* Check values #4 */
SELECT tableoid::REGCLASS, *
FROM test_update_trigger.test_range
WHERE val = -1
ORDER BY comment;

SELECT count(*) FROM test_update_trigger.test_range;


/* Update non-key column */
UPDATE test_update_trigger.test_range SET comment = 'test!' WHERE val = 100;

/* Check values #5 */
SELECT tableoid::REGCLASS, *
FROM test_update_trigger.test_range
WHERE val = 100
ORDER BY comment;

SELECT count(*) FROM test_update_trigger.test_range;


/* Try moving row into a gap (ERROR) */
DROP TABLE gogudb_partition_table._test_update_trigger_4_test_range;
UPDATE test_update_trigger.test_range SET val = 35 WHERE val = 70;

/* Check values #6 */
SELECT tableoid::REGCLASS, *
FROM test_update_trigger.test_range
WHERE val = 70
ORDER BY comment;

SELECT count(*) FROM test_update_trigger.test_range;


/* Test trivial move (same key) */
UPDATE test_update_trigger.test_range SET val = 65 WHERE val = 65;

/* Check values #7 */
SELECT tableoid::REGCLASS, *
FROM test_update_trigger.test_range
WHERE val = 65
ORDER BY comment;

SELECT count(*) FROM test_update_trigger.test_range;


/* Test tuple conversion (attached partition) */
CREATE TABLE test_update_trigger.test_range_inv(comment TEXT, val NUMERIC NOT NULL);
SELECT attach_range_partition('test_update_trigger.test_range',
							  'test_update_trigger.test_range_inv',
							  101::NUMERIC, 111::NUMERIC);
UPDATE test_update_trigger.test_range SET val = 105 WHERE val = 60;

/* Check values #8 */
SELECT tableoid::REGCLASS, *
FROM test_update_trigger.test_range
WHERE val = 105
ORDER BY comment;

SELECT count(*) FROM test_update_trigger.test_range;


/* Test tuple conversion (dropped column) */
ALTER TABLE test_update_trigger.test_range DROP COLUMN comment CASCADE;
SELECT append_range_partition('test_update_trigger.test_range');
UPDATE test_update_trigger.test_range SET val = 115 WHERE val = 55;

/* Check values #9 */
SELECT tableoid::REGCLASS, *
FROM test_update_trigger.test_range
WHERE val = 115;

SELECT count(*) FROM test_update_trigger.test_range;



/* Partition table by HASH (INT4) */
CREATE TABLE test_update_trigger.test_hash(val INT4 NOT NULL, comment TEXT);
INSERT INTO test_update_trigger.test_hash SELECT i, i FROM generate_series(1, 10) i;
SELECT create_hash_partitions('test_update_trigger.test_hash', 'val', 3);
SELECT create_update_triggers('test_update_trigger.test_hash');


/* Move all rows into single partition */
UPDATE test_update_trigger.test_hash SET val = 1;

/* Check values #1 */
SELECT tableoid::REGCLASS, *
FROM test_update_trigger.test_hash
WHERE val = 1
ORDER BY comment;

SELECT count(*) FROM test_update_trigger.test_hash;


/* Don't move any rows */
UPDATE test_update_trigger.test_hash SET val = 3 WHERE val = 2;

/* Check values #2 */
SELECT tableoid::REGCLASS, *
FROM test_update_trigger.test_hash
WHERE val = 3
ORDER BY comment;

SELECT count(*) FROM test_update_trigger.test_hash;



DROP SCHEMA test_update_trigger CASCADE;
DROP EXTENSION gogudb;

