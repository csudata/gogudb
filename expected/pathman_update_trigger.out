\set VERBOSITY terse
SET search_path = 'public','_gogu';
CREATE EXTENSION gogudb;
CREATE SCHEMA test_update_trigger;
/* Partition table by RANGE (NUMERIC) */
CREATE TABLE test_update_trigger.test_range(val NUMERIC NOT NULL, comment TEXT);
INSERT INTO test_update_trigger.test_range SELECT i, i FROM generate_series(1, 100) i;
SELECT create_range_partitions('test_update_trigger.test_range', 'val', 1, 10);
 create_range_partitions 
-------------------------
                      10
(1 row)

SELECT create_update_triggers('test_update_trigger.test_range');
 create_update_triggers 
------------------------
 
(1 row)

/* Update values in 1st partition (rows remain there) */
UPDATE test_update_trigger.test_range SET val = 5 WHERE val <= 10;
/* Check values #1 */
SELECT tableoid::REGCLASS, *
FROM test_update_trigger.test_range
WHERE val < 10
ORDER BY comment;
                         tableoid                         | val | comment 
----------------------------------------------------------+-----+---------
 gogudb_partition_table._test_update_trigger_1_test_range |   5 | 1
 gogudb_partition_table._test_update_trigger_1_test_range |   5 | 10
 gogudb_partition_table._test_update_trigger_1_test_range |   5 | 2
 gogudb_partition_table._test_update_trigger_1_test_range |   5 | 3
 gogudb_partition_table._test_update_trigger_1_test_range |   5 | 4
 gogudb_partition_table._test_update_trigger_1_test_range |   5 | 5
 gogudb_partition_table._test_update_trigger_1_test_range |   5 | 6
 gogudb_partition_table._test_update_trigger_1_test_range |   5 | 7
 gogudb_partition_table._test_update_trigger_1_test_range |   5 | 8
 gogudb_partition_table._test_update_trigger_1_test_range |   5 | 9
(10 rows)

SELECT count(*) FROM test_update_trigger.test_range;
 count 
-------
   100
(1 row)

/* Update values in 2nd partition (rows move to 3rd partition) */
UPDATE test_update_trigger.test_range SET val = val + 10 WHERE val > 10 AND val <= 20;
/* Check values #2 */
SELECT tableoid::REGCLASS, *
FROM test_update_trigger.test_range
WHERE val > 20 AND val <= 30
ORDER BY comment;
                         tableoid                         | val | comment 
----------------------------------------------------------+-----+---------
 gogudb_partition_table._test_update_trigger_3_test_range |  21 | 11
 gogudb_partition_table._test_update_trigger_3_test_range |  22 | 12
 gogudb_partition_table._test_update_trigger_3_test_range |  23 | 13
 gogudb_partition_table._test_update_trigger_3_test_range |  24 | 14
 gogudb_partition_table._test_update_trigger_3_test_range |  25 | 15
 gogudb_partition_table._test_update_trigger_3_test_range |  26 | 16
 gogudb_partition_table._test_update_trigger_3_test_range |  27 | 17
 gogudb_partition_table._test_update_trigger_3_test_range |  28 | 18
 gogudb_partition_table._test_update_trigger_3_test_range |  29 | 19
 gogudb_partition_table._test_update_trigger_3_test_range |  30 | 20
 gogudb_partition_table._test_update_trigger_3_test_range |  21 | 21
 gogudb_partition_table._test_update_trigger_3_test_range |  22 | 22
 gogudb_partition_table._test_update_trigger_3_test_range |  23 | 23
 gogudb_partition_table._test_update_trigger_3_test_range |  24 | 24
 gogudb_partition_table._test_update_trigger_3_test_range |  25 | 25
 gogudb_partition_table._test_update_trigger_3_test_range |  26 | 26
 gogudb_partition_table._test_update_trigger_3_test_range |  27 | 27
 gogudb_partition_table._test_update_trigger_3_test_range |  28 | 28
 gogudb_partition_table._test_update_trigger_3_test_range |  29 | 29
 gogudb_partition_table._test_update_trigger_3_test_range |  30 | 30
(20 rows)

SELECT count(*) FROM test_update_trigger.test_range;
 count 
-------
   100
(1 row)

/* Move single row */
UPDATE test_update_trigger.test_range SET val = 90 WHERE val = 80;
/* Check values #3 */
SELECT tableoid::REGCLASS, *
FROM test_update_trigger.test_range
WHERE val = 90
ORDER BY comment;
                         tableoid                         | val | comment 
----------------------------------------------------------+-----+---------
 gogudb_partition_table._test_update_trigger_9_test_range |  90 | 80
 gogudb_partition_table._test_update_trigger_9_test_range |  90 | 90
(2 rows)

SELECT count(*) FROM test_update_trigger.test_range;
 count 
-------
   100
(1 row)

/* Move single row (create new partition) */
UPDATE test_update_trigger.test_range SET val = -1 WHERE val = 50;
/* Check values #4 */
SELECT tableoid::REGCLASS, *
FROM test_update_trigger.test_range
WHERE val = -1
ORDER BY comment;
                         tableoid                          | val | comment 
-----------------------------------------------------------+-----+---------
 gogudb_partition_table._test_update_trigger_11_test_range |  -1 | 50
(1 row)

SELECT count(*) FROM test_update_trigger.test_range;
 count 
-------
   100
(1 row)

/* Update non-key column */
UPDATE test_update_trigger.test_range SET comment = 'test!' WHERE val = 100;
/* Check values #5 */
SELECT tableoid::REGCLASS, *
FROM test_update_trigger.test_range
WHERE val = 100
ORDER BY comment;
                         tableoid                          | val | comment 
-----------------------------------------------------------+-----+---------
 gogudb_partition_table._test_update_trigger_10_test_range | 100 | test!
(1 row)

SELECT count(*) FROM test_update_trigger.test_range;
 count 
-------
   100
(1 row)

/* Try moving row into a gap (ERROR) */
DROP TABLE gogudb_partition_table._test_update_trigger_4_test_range;
UPDATE test_update_trigger.test_range SET val = 35 WHERE val = 70;
ERROR:  cannot spawn a partition
/* Check values #6 */
SELECT tableoid::REGCLASS, *
FROM test_update_trigger.test_range
WHERE val = 70
ORDER BY comment;
                         tableoid                         | val | comment 
----------------------------------------------------------+-----+---------
 gogudb_partition_table._test_update_trigger_7_test_range |  70 | 70
(1 row)

SELECT count(*) FROM test_update_trigger.test_range;
 count 
-------
    90
(1 row)

/* Test trivial move (same key) */
UPDATE test_update_trigger.test_range SET val = 65 WHERE val = 65;
/* Check values #7 */
SELECT tableoid::REGCLASS, *
FROM test_update_trigger.test_range
WHERE val = 65
ORDER BY comment;
                         tableoid                         | val | comment 
----------------------------------------------------------+-----+---------
 gogudb_partition_table._test_update_trigger_7_test_range |  65 | 65
(1 row)

SELECT count(*) FROM test_update_trigger.test_range;
 count 
-------
    90
(1 row)

/* Test tuple conversion (attached partition) */
CREATE TABLE test_update_trigger.test_range_inv(comment TEXT, val NUMERIC NOT NULL);
SELECT attach_range_partition('test_update_trigger.test_range',
							  'test_update_trigger.test_range_inv',
							  101::NUMERIC, 111::NUMERIC);
       attach_range_partition       
------------------------------------
 test_update_trigger.test_range_inv
(1 row)

UPDATE test_update_trigger.test_range SET val = 105 WHERE val = 60;
/* Check values #8 */
SELECT tableoid::REGCLASS, *
FROM test_update_trigger.test_range
WHERE val = 105
ORDER BY comment;
              tableoid              | val | comment 
------------------------------------+-----+---------
 test_update_trigger.test_range_inv | 105 | 60
(1 row)

SELECT count(*) FROM test_update_trigger.test_range;
 count 
-------
    90
(1 row)

/* Test tuple conversion (dropped column) */
ALTER TABLE test_update_trigger.test_range DROP COLUMN comment CASCADE;
SELECT append_range_partition('test_update_trigger.test_range');
                  append_range_partition                   
-----------------------------------------------------------
 gogudb_partition_table._test_update_trigger_12_test_range
(1 row)

UPDATE test_update_trigger.test_range SET val = 115 WHERE val = 55;
/* Check values #9 */
SELECT tableoid::REGCLASS, *
FROM test_update_trigger.test_range
WHERE val = 115;
                         tableoid                          | val 
-----------------------------------------------------------+-----
 gogudb_partition_table._test_update_trigger_12_test_range | 115
(1 row)

SELECT count(*) FROM test_update_trigger.test_range;
 count 
-------
    90
(1 row)

/* Partition table by HASH (INT4) */
CREATE TABLE test_update_trigger.test_hash(val INT4 NOT NULL, comment TEXT);
INSERT INTO test_update_trigger.test_hash SELECT i, i FROM generate_series(1, 10) i;
SELECT create_hash_partitions('test_update_trigger.test_hash', 'val', 3);
 create_hash_partitions 
------------------------
                      3
(1 row)

SELECT create_update_triggers('test_update_trigger.test_hash');
 create_update_triggers 
------------------------
 
(1 row)

/* Move all rows into single partition */
UPDATE test_update_trigger.test_hash SET val = 1;
/* Check values #1 */
SELECT tableoid::REGCLASS, *
FROM test_update_trigger.test_hash
WHERE val = 1
ORDER BY comment;
                        tableoid                         | val | comment 
---------------------------------------------------------+-----+---------
 gogudb_partition_table._test_update_trigger_1_test_hash |   1 | 1
 gogudb_partition_table._test_update_trigger_1_test_hash |   1 | 10
 gogudb_partition_table._test_update_trigger_1_test_hash |   1 | 2
 gogudb_partition_table._test_update_trigger_1_test_hash |   1 | 3
 gogudb_partition_table._test_update_trigger_1_test_hash |   1 | 4
 gogudb_partition_table._test_update_trigger_1_test_hash |   1 | 5
 gogudb_partition_table._test_update_trigger_1_test_hash |   1 | 6
 gogudb_partition_table._test_update_trigger_1_test_hash |   1 | 7
 gogudb_partition_table._test_update_trigger_1_test_hash |   1 | 8
 gogudb_partition_table._test_update_trigger_1_test_hash |   1 | 9
(10 rows)

SELECT count(*) FROM test_update_trigger.test_hash;
 count 
-------
    10
(1 row)

/* Don't move any rows */
UPDATE test_update_trigger.test_hash SET val = 3 WHERE val = 2;
/* Check values #2 */
SELECT tableoid::REGCLASS, *
FROM test_update_trigger.test_hash
WHERE val = 3
ORDER BY comment;
 tableoid | val | comment 
----------+-----+---------
(0 rows)

SELECT count(*) FROM test_update_trigger.test_hash;
 count 
-------
    10
(1 row)

DROP SCHEMA test_update_trigger CASCADE;
NOTICE:  drop cascades to 18 other objects
DROP EXTENSION gogudb;
